Enable use of alternate hardware UARTs for logging

brandond opened the pull request:
## Description:
Enable use of Serial1 on ESP8266 and Serial1/Serial2 on ESP32 for logging.
This is frequently done on ESP8266 to allow use of Serial for UART TX+RX,
while  maintaining logging output on Serial1 which is TX-only via GPIO2.

**Pull request in [esphomeyaml](https://github.com/OttoWinter/esphomeyaml) with YAML changes (if applicable):** OttoWinter/esphomeyaml#427
**Pull request in [esphomedocs](https://github.com/OttoWinter/esphomedocs) with documentation (if applicable):** OttoWinter/esphomedocs#161

## Checklist:
  - [X] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphomedocs](https://github.com/OttoWinter/esphomedocs).

brandond committed:
Enable use of alternate hardware UARTs for logging

Enable use of Serial1 on ESP8266 and Serial1/Serial2 on ESP32 for logging.
This is frequently done on ESP8266 to allow use of Serial for UART TX+RX,
while  maintaining logging output on Serial1 which is TX-only via GPIO2.

OttoWinter commented on src/esphomelib/uart_component.cpp:
```diff
@@ -22,15 +22,20 @@ UARTComponent::UARTComponent(int8_t tx_pin, int8_t rx_pin, uint32_t baud_rate)
 #ifdef ARDUINO_ARCH_ESP8266
   if (this->rx_pin_ == 3 && this->tx_pin_ == 1) {
     this->hw_serial_ = &Serial;
+  } elif (this->rx_pin_ == 8 && this->tx_pin_ == 2) {
```
It looks like the ESP8266 does not support RX on the UART1 interface (see https://github.com/esp8266/Arduino/blob/master/cores/esp8266/uart.cpp#L631-L637).

brandond replied:
UART1 supports RX, it's just that the pin (IO8) is also used for SDIO_DATA1 and is disabled when SPI is in use - which is basically always.

UART0 is also interesting in that it supports RTS/CTS on IO13/IO12, and RTS/CTS and RX/TX can be swapped. It's not anywhere near the flexibility of ESP32, but it's not bad given the overall low pinout of the device.

OttoWinter commented on src/esphomelib/uart_component.cpp:
```diff
@@ -22,15 +22,20 @@ UARTComponent::UARTComponent(int8_t tx_pin, int8_t rx_pin, uint32_t baud_rate)
 #ifdef ARDUINO_ARCH_ESP8266
   if (this->rx_pin_ == 3 && this->tx_pin_ == 1) {
     this->hw_serial_ = &Serial;
+  } elif (this->rx_pin_ == 8 && this->tx_pin_ == 2) {
```
`elif` -> `else if`

OttoWinter reviewed (commented):
While reviewing this PR I realized some of the logic in the uart component was a bit funky :)

For example, the HW serial should be used to if either one of RX or TX pin matches (or both). Until now the HW serial was only used on the ESP8266 iff tx&rx were used.

Created https://github.com/esphome/esphome-core/pull/487 to fix this. I think this PR can be rebased on that once that's complete.

brandond commented on src/esphomelib/uart_component.cpp:
```diff
@@ -22,15 +22,20 @@ UARTComponent::UARTComponent(int8_t tx_pin, int8_t rx_pin, uint32_t baud_rate)
+  } elif (this->rx_pin_ == 9 && this->tx_pin_ == 10) {
+    this->hw_serial_ = &Serial1
+  } elif (this->rx_pin_ == 16 && this->tx_pin_ == 17) {
+    this->hw_serial_ = &Serial2
```
Sure, but UART1 and UART2 are set up on these pins by default by the Arduino framework. I'm not sure what happens if someone tries to re-use those same pins on another UART - I guess nothing, as long as they don't also try to use the Arduino global Serial instances that are preconfigured to use those pins?

brandond commented:
I just rebased on top of your fixes. I opted to keep the Serial1/Serial2 reuse stuff since I think it might be useful. I'll put in a docs PR to describe how to make use of it.

brandond commented:
@OttoWinter can you sanity-check my approach here? I'd kind of like to keep it simple and allow users to move the logger over to another hardware UART. However, since there's a full UART component now, it feels like the logger should really use that instead of just using the raw global Serial object. We'd need to expose some of the low-level hardware capabilities (`setDebugOutput`, `swap`, etc), through UART, but that's not terrible. 

That would change the logger config to something like this:
```
uart:
  - tx_pin: 1
    rx_pin: 3
    baud_rate: 9600
    id: uart0
  - tx_pin: 2
    rx_pin: 8
    baud_rate: 115200
    id: uart1

logger:
  level: DEBUG
  id: logger
  #baud_rate: 115200 # Exclusive of uart_id
  uart_id: uart1
```

This would also mean that UARTs need to be set up before the logger; I'm not sure if that will break anything else?

OttoWinter commented:
Yeah that would be an option. On the ESP32 that would add any functionality. On the ESP8266 it would enable the use of software serial for sending.

However, I don't think it's worth the effort TBH. As you said, the UART component would need to be set up before the logger, and a whole bunch of edge cases would need to be handled.

Then there's the problem of speed. The logger is called often and should not do much work, software serial on ESP8266 would probably make any log call quite slow. Plus the logger should never break - the logger is the first thing that's initialized, if that breaks the user won't:
* get any logs showing what went wrong
* even get OTA safe mode.

And then there's the question why we should do this. Sure it would maybe be cleaner on a high level with more of a plug-and-play nature, but I doubt any user would ever use the added functionality that offers.

brandond commented:
I really just need to get the logger off to UART1 on esp8266 so I can free up UART0 for the communication with the coprocessor on #306 while not loosing the serial log capability. This is what the stock Tuya firmware does.

brandond commented:
OK, I think this is a fairly simple approach that lets me accomplish what I'm trying to do without complicating things too much.

brandond committed:
Fix typo

OttoWinter reviewed (approved):
👍

OttoWinter commented on src/esphomelib/log_component.h:
```diff
@@ -17,6 +17,22 @@
+ * Advanced configuration (pin selection, etc) is not supported.
+ */
+enum UARTSelection {
+  ESPHOMELIB_UART0 = 0,
```
For enums, I prefer to prefix them with the enum name (I think this is google style too, but not sure).

So `UART_SELECTION_UART0 = 0,`

(Or in this case mostly just so that I don't have to change this to `ESPHOME_UART0` in https://github.com/esphome/esphome-core/pull/494 again 😜 )

brandond replied:
Can do! I think I copied it from the log stuff - ESPHOMELIB_LOG_X etc.

brandond committed:
ESPHOMELIB_UART -> UART_SELECTION

brandond commented:
I finally got a chance to test this on real hardware, and it mostly looks good. There's some Arduino stuff that's hardcoded to output on a specific port using the DEBUG_ESP_PORT macro; I'll have to add a note to the docs about that.

brandond committed:
Disable debugging output when UART component takes over

OttoWinter commented on src/esphomelib/uart_component.cpp:
```diff
@@ -115,11 +120,22 @@ void UARTComponent::flush() {
-    // for example if RX disabled but TX pin is 1 we still want to use Serial.
     this->hw_serial_ = &Serial;
     this->hw_serial_->begin(this->baud_rate_);
+    this->hw_serial_->setDebugOutput(false);
```
I do not think we should disable debug output here. If the user doesn't want debug output on a serial port they should disable the logger component for it. 

If they don't, even though we disable the SDKs debug output here, the user will still get debug output through esphome's own logging engine.

brandond replied:
I'll re-test without this but with the platformio build flag. My experience was that I still got debug output on UART0 unless I called setDebugOutput(false) on it here. This is due to the Arduino framework enabling it by default even before the esphome logger takes over. The esphome logger calls setDebugOutput(true) on the uart selected for logging, so the change is just to put debugging on the logger UART, not to disable it entirely.

brandond replied:
I guess that in the case of UART and Logger sharing a port, the UART would disable it after the Logger enabled it, so you're right - this isn't safe to do unconditionally.

brandond committed:
Don't disable debug output; may be shared with logger.

Setting the debug serial port via build flags should resolve
the issue I was attempting to work around with this fix anyway.

OttoWinter reviewed (approved):
Ok, this looks great now! Thanks!
